### PR TITLE
Outline polling loop for Permalink service 

### PR DESCRIPTION
**Pull request by rexim**
Split #20 

---

### TODOs
- `1cdf1e3c-d19b-493d-9d33-f88c92ee24e3`
- `83b40d5c-6af7-4268-a2a2-960d051102ae`
- `19e120b5-c7a8-426e-b54c-170cf1d7d3da`
- `e2419ff0-47b9-4a54-91db-979b84de030e`

### CODE REVIEW

File: docs/Permalink.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+### Description
```
Comment (rexim) [inline docs/Permalink.md]: Replace that with `# Permalink`.

File: docs/Permalink.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+```
+https://github.com/tsoding/Issuestant/blob/master/src/main/scala/me/rexim/Mixer.scala
+```
+
+should be replaced with
+
+```
+https://github.com/tsoding/Issuestant/blob/<latest-commit-hash>/src/main/scala/me/rexim/Mixer.scala
+```
```
Comment (rexim) [inline docs/Permalink.md]: Add information about requesting appropriate permissions

File: src/main/scala/me/rexim/issuestant/Main.scala [hunk 3]
```diff
@@ -1,13 +1,32 @@
...
+import org.http4s.client.blaze._
 
 object Main extends ServerApp {
+  def pollingLoop(permalink: Permalink): Unit = {
+    permalink.update.flatMap { nextPermalink =>
+      Task.schedule(pollingLoop(nextPermalink), FiniteDuration(1, SECONDS))
+    }.runAsync { _ => }
+  }
+
   // TODO: Use the HTTP port passed by heroku
```
Comment (rexim) [inline src/main/scala/me/rexim/issuestant/Main.scala]: Link that with the corresponding issue

File: src/main/scala/me/rexim/issuestant/Permalink.scala [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+// TODO(1cdf1e3c-d19b-493d-9d33-f88c92ee24e3): Implement Permalink
+
+/** Issuestant service for Permalink protocol (see doc/Permalink.md)
+  *
+  * The class is stateless. The next state is returned by update()
+  * method. The state should be managed by the invoker.
+  *
+  * @param eventSource the source of GitHub events
+  */
+class Permalink(eventSource: EventsSource) {
```
Comment (rexim) [inline src/main/scala/me/rexim/issuestant/Permalink.scala]: Your assumptions are wrong. The Permalink service also depends on entities that provide ability to modify comments, issue descriptions and file issues.

**Comment by rexim** [comment]
@rexim see how to properly deal with the coverage threshold when we have unfinished classes.

**Comment by rexim** [comment]
> @rexim see how to properly deal with the coverage threshold when we have unfinished classes.

We can actually try to exclude them from coverage. It's similar to the Unit Test of unfinished classes. Such Unit Tests are disable until the class is implemented.

See https://github.com/scoverage/sbt-scoverage#exclude-classes-and-packages

File: src/main/scala/me/rexim/issuestant/Permalink.scala (hunk 4 above)
Comment (rexim) [inline src/main/scala/me/rexim/issuestant/Permalink.scala]: Disable coverage for this class. See https://github.com/scoverage/sbt-scoverage#exclude-classes-and-packages

File: src/main/scala/me/rexim/issuestant/github/EventsSource.scala [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+/** The source of GitHub events
+  *
+  * The class is stateless. The next state is returned by nextEvents()
+  * method. The state should be managed by the invoker.
+  *
+  * @param client an http client for requesting GitHub API
+  * @param owner the owner of the repo
+  * @param repo the name of the repo
+  */
+class EventsSource (client: Client, owner: String, repo: String) {
```
Comment (rexim) [inline src/main/scala/me/rexim/issuestant/github/EventsSource.scala]: Disable coverage for this class https://github.com/scoverage/sbt-scoverage#exclude-classes-and-packages

File: docs/Permalink.md (hunk 1 above)
Comment (rexim) [inline docs/Permalink.md]: Done

File: docs/Permalink.md (hunk 2 above)
Comment (rexim) [inline docs/Permalink.md]: Done

File: src/main/scala/me/rexim/issuestant/Main.scala (hunk 3 above)
Comment (rexim) [inline src/main/scala/me/rexim/issuestant/Main.scala]: Done

File: src/main/scala/me/rexim/issuestant/github/EventsSource.scala (hunk 5 above)
Comment (rexim) [inline src/main/scala/me/rexim/issuestant/github/EventsSource.scala]: Done

File: src/main/scala/me/rexim/issuestant/Main.scala [hunk 6]
```diff
@@ -1,13 +1,32 @@
...
-  // TODO: Use the HTTP port passed by heroku
+  def pollingLoop(permalink: Permalink): Unit = {
+    permalink.update.flatMap { nextPermalink =>
+      Task.schedule(pollingLoop(nextPermalink), FiniteDuration(1, SECONDS))
+    }.runAsync { _ => }
+  }
+
+  // TODO(19f86d8b-7c7f-4acc-af5b-a3a4c29cec5c): Use the HTTP port passed by heroku
   override def server(args: List[String]): Task[Server] = {
+    Task.schedule(pollingLoop(new Permalink(
```
Comment (rexim) [inline src/main/scala/me/rexim/issuestant/Main.scala]: Extract this polling loop into a separate class. That will help you with testing it and meeting the coverage threshold.

File: src/main/scala/me/rexim/issuestant/Main.scala [hunk 7]
```diff
@@ -1,13 +1,32 @@
...
 
 import org.http4s.server.{Server, ServerApp}
 import org.http4s.server.blaze._
+import org.http4s.client.blaze._
 
 object Main extends ServerApp {
-  // TODO: Use the HTTP port passed by heroku
+  def pollingLoop(permalink: Permalink): Unit = {
+    permalink.update.flatMap { nextPermalink =>
+      Task.schedule(pollingLoop(nextPermalink), FiniteDuration(1, SECONDS))
```
Comment (ForNeVeR) [inline src/main/scala/me/rexim/issuestant/Main.scala]: I suggest to extract that logic into another function that'll handle exceptions etc. properly. I don't like that `pollingLoop` is responsible for two things:

- a. Processing the business logic.
- b. Maintaining the async loop.

_b._ is a technical thing and could be extracted and properly tested isolated from the business logic. We're possibly going to have many "polling loops" like that for various thingies.

Comment (rexim) [inline src/main/scala/me/rexim/issuestant/Main.scala]: Yeah, I completely agree. I'm in the middle of doing exactly that. :D I'm also research how we can apply scalaz-streams here. Looks like a thing that can fit our purposes. Will update shortly on that.

Thanks for the remark!

File: src/main/scala/me/rexim/issuestant/Main.scala (hunk 6 above)
Comment (rexim) [inline src/main/scala/me/rexim/issuestant/Main.scala]: Done

**Comment by rexim** [comment]
I'm done on my remarks. If anybody who's interested in helping has anything else please let me know here. I'm gonna wait until tomorrow.

**Comment by rexim** [comment]
@ForNeVeR thanks for the review!

**Pull request merged**